File: app/channels/chat_channel.rb
class ChatChannel < ApplicationCable::Channel
  def subscribed
    # stream_from "some_channel"
    stream_from "chat_#{params[:room]}"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def send_message(data)
    message = Message.create!(user: data['user'], content: data['message'])
    ActionCable.server.broadcast("chat_#{params[:room]}", message: render_message(message))
  end

  private

  def render_message(message)
    ApplicationController.renderer.render(partial: 'messages/message', locals: { message: message })
  end
end

File: app/javascript/channels/chat_channel.js
import consumer from "./consumer"

const room = "general"  // You can customize this based on room parameters

consumer.subscriptions.create({ channel: "ChatChannel", room: room }, {
  received(data) {
    // Append the received message to the chat window
    const messagesContainer = document.getElementById('messages')
    messagesContainer.innerHTML += data.message
  },

  sendMessage(user, message) {
    this.perform('send_message', { user: user, message: message });
  }
});
